zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ZooKeeperUpdatingListener.java

```diff
@@ -30,68 +37,97 @@
                                      Endpoint endpoint) {
-        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
+        checkArgument(!Strings.isNullOrEmpty(zkConnectionStr), "zkConnectionStr");
+        ExponentialBackoffRetry retryPolicy = new ExponentialBackoffRetry(1000, 3);
+        client = CuratorFrameworkFactory.builder()
+                .connectString(zkConnectionStr)
+                .retryPolicy(retryPolicy)
+                .sessionTimeoutMs(sessionTimeout)
+                .build();
         this.zNodePath = requireNonNull(zNodePath, "zNodePath");
-        this.endpoint = requireNonNull(endpoint, "endpoint");
-        this.sessionTimeout = sessionTimeout;
+        this.nodeValueCodec = NodeValueCodec.DEFAULT;
+        this.endpoint = endpoint;
     }
 
     /**
      * A ZooKeeper server listener, used for register server into ZooKeeper.
+     *
+     * @deprecated Use {@link ZooKeeperUpdatingListenerBuilder}.
+     *
      * @param zkConnectionStr ZooKeeper connection string
      * @param zNodePath       ZooKeeper node path(under which this server will registered)
      * @param sessionTimeout  session timeout
      */
+    @Deprecated
     public ZooKeeperUpdatingListener(String zkConnectionStr, String zNodePath, int sessionTimeout) {
-        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
-        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
-        this.sessionTimeout = sessionTimeout;
+        this(zkConnectionStr, zNodePath, sessionTimeout, null);
     }
 
     @Override
     public void serverStarted(Server server) {
         if (endpoint == null) {
             assert server.activePort().isPresent();
             endpoint = Endpoint.of(server.defaultHostname(),
-                                   server.activePort().get()
-                                         .localAddress().getPort());
+                    server.activePort().get()
+                            .localAddress().getPort());
+        }
+        client.start();
+        String key = endpoint.host() + '_' + endpoint.port();
+        byte[] value = nodeValueCodec.encode(endpoint);
+        try {
+            client.create()
+                    .creatingParentsIfNeeded()
+                    .withMode(CreateMode.EPHEMERAL)
+                    .forPath(zNodePath + '/' + key, value);
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
         }
-        connector = new ZooKeeperRegistration(zkConnectionStr, zNodePath, sessionTimeout, endpoint);
     }
 
     @Override
     public void serverStopping(Server server) {
-        if (connector != null) {
-            connector.close(true);
-        }
-    }
-
-    @Nullable
-    @VisibleForTesting
-    ZooKeeperRegistration getConnector() {
-        return connector;
-    }
-
-    @Nullable
-    @VisibleForTesting
-    Endpoint getEndpoint() {
-        return endpoint;
+        client.close();
```

Comment:
Do not close `CuratorFramework` but only `PathChildrenCache` if a user created this listener by specifying an existing `CuratorFramework` instance.